app/src/main/java/com/celzero/bravedns/service/ConnectionMonitor.kt

```diff
@@ -0,0 +1,240 @@
+import android.util.Log
+import com.celzero.bravedns.ui.HomeScreenActivity.GlobalVariable.DEBUG
+import com.celzero.bravedns.util.Constants.Companion.LOG_TAG
+import com.google.common.collect.Sets
+import org.koin.core.component.KoinComponent
+import org.koin.core.component.inject
+
+
+class ConnectionMonitor(context: Context, val networkListener: NetworkListener) : ConnectivityManager.NetworkCallback(), KoinComponent {
+    private val networkRequest: NetworkRequest = NetworkRequest.Builder().addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET).build()
+
+    // An Android handler thread internally operates on a looper.
+    // Defined as @Volatile so that the looper in the Handler thread so that
+    // the value loaded from memory instead of Thread's local cache.
+    // ref - https://alvinalexander.com/java/jwarehouse/android/core/java/android/app/IntentService.java.shtml
+    @Volatile private var handlerThread: HandlerThread? = null
+    private var serviceHandler: NetworkRequestHandler? = null
+    private val persistentState by inject<PersistentState>()
+
+    // ref - https://developer.android.com/reference/kotlin/java/util/LinkedHashSet
+    // The network list is maintained in a linked-hash-set to preserve insertion and iteration
+    // order. This is required because {@link android.net.VpnService#setUnderlyingNetworks}
+    // defines network priority depending on the iteration order, that is, the network
+    // in the 0th index is preferred over the one at 1st index, and so on.
+    var currentNetworks: LinkedHashSet<Network> = linkedSetOf()
+    var currentActiveNetwork: Network? = null
+
+    var connectivityManager: ConnectivityManager = context.applicationContext.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
+
+
+    companion object {
+        // add active network as underlying vpn network
+        const val MSG_ADD_ACTIVE_NETWORK = 1
+        // add all available networks as underlying vpn networks
+        const val MSG_ADD_ALL_NETWORKS = 2
+    }
+
+    init {
+        connectivityManager.registerNetworkCallback(networkRequest, this)
+        this.handlerThread = HandlerThread("NetworkRequestHandler.HandlerThread")
+        this.handlerThread?.start()
+        this.serviceHandler = NetworkRequestHandler(handlerThread?.looper, this)
+    }
+
+    interface NetworkListener {
+        fun onNetworkDisconnected()
+        fun onNetworkConnected(networks: LinkedHashSet<Network>?)
+    }
+
+    fun removeCallBack() {
+        connectivityManager.unregisterNetworkCallback(this)
+        destroy()
+    }
+
+    private fun destroy() {
+        this.serviceHandler?.removeCallbacksAndMessages(null)
+        this.handlerThread?.quitSafely()
+        this.handlerThread = null
+        this.serviceHandler = null
+    }
+
+    override fun onAvailable(network: Network) {
+        if(DEBUG) Log.d(LOG_TAG, "ConnectionMonitor - onAvailable")
+        handleNetworkChange(isForceUpdate = false)
```

Comment:
nit: why not rely on argument default value set to `false` instead? Ref: https://www.baeldung.com/kotlin/default-named-arguments#1-default-arguments-in-kotlin

Reply:
Added default value for the argument.

Reply:
Changed